pkg/k8s/cninode_cleanup.go

```diff
@@ -0,0 +1,119 @@
+// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//     http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package k8s
+
+import (
+	"context"
+	"time"
+
+	"github.com/aws/amazon-vpc-resource-controller-k8s/apis/vpcresources/v1alpha1"
+	"github.com/aws/amazon-vpc-resource-controller-k8s/pkg/config"
+	rcHealthz "github.com/aws/amazon-vpc-resource-controller-k8s/pkg/healthz"
+
+	"github.com/go-logr/logr"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/healthz"
+)
+
+type CNINodeCleaner struct {
+	Log            logr.Logger
+	shutdown       bool
+	ctx            context.Context
+	k8sClient      client.Client
+	leakedCNINodes []string
+}
+
+func (e *CNINodeCleaner) SetupWithManager(ctx context.Context, mgr ctrl.Manager, healthzHandler *rcHealthz.HealthzHandler) error {
+	e.ctx = ctx
+	e.k8sClient = mgr.GetClient()
+	healthzHandler.AddControllersHealthCheckers(
+		map[string]healthz.Checker{
+			"health-cninode-cleaner": rcHealthz.SimplePing("cninode cleanup", e.Log),
+		},
+	)
+
+	return mgr.Add(e)
+}
+
+// StartCNINodeCleaner starts the CNINOde Cleaner routine that cleans up dangling CNINode created by the controller
+func (e *CNINodeCleaner) Start(ctx context.Context) error {
```

Comment:
Can we use OwnerReferences to handle garbage collection, instead of building a custom controller for this?

Reply:
We are already using the OwnerReference for CNINode CR. I am just paranoid on possible leakage for any reason k8s couldn't delete the CNINodes... That is also why I set this for every 12 hours.

Reply:
> I am just paranoid on possible leakage for any reason

I'd be very cautious of this line of reasoning. The kubernetes garbage collector is a core mechanism that drives things like pod deletion when replicasets are deleted -- if we can't rely on it, Kubernetes will break in many ways. We should keep systems as simple as possible, and test them thoroughly to have confidence that the mechanisms are working as intended.

Reply:
Good point. Relying on it and trusting it. Ack. I will remove the extra cleanup.